rust/template/differential_datalog/ddval.rs

```diff
@@ -0,0 +1,423 @@
+//! DDValue: Generic value type stored in all differential-dataflow relations.
+//!
+//! Rationale: Differential dataflow allows the user to assign an arbitrary user-defined type to
+//! each collection.  It relies on Rust's static dispatch mechanism to specialize its internal
+//! machinery for each user-defined type.  Unfortunately, beyond very simple programs this leads to
+//! extremely long compilation times.  One workaround that we used to rely on is to declare a
+//! single enum type with a variant per concrete type used in at least one relation.  This make
+//! compilation feasible, but still very slow (~6 minutes for a simple DDlog program and ~10
+//! minutes for complex programs).
+//!
+//! Another alternative we implement here is to use a fixed value type that does not depend on
+//! a concrete DDlog program and rely on dynamic dispatch to forward operations that DD expects
+//! all values to implement (comparison, hashing, etc.) to their concrete implementations.  This
+//! way this crate (differetial-datalog) can be compiled all the way to binary code separately
+//! from the DDlog program using it and does not need to be re-compiled when the DDlog program
+//! changes.  Thus, the only part that must be re-compiled on changes to the DDlog code is the
+//! auto-generated crate that declares concrete value types and rules.  This is much faster than
+//! re-compiling both crates together.
```

Comment:
Not until we move the shared crate out of the template and in a separate location (even then I wonder what it will take to convince Rust not to re-compile it for each test).